### PR TITLE
Perform verification of `SelectionCollateralError` values.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue Number

ADP-1037

## Summary

This PR adds verification of `SelectionCollateralError` values.

We verify that:
- The reported `largestCombinationAvailable` value does not include any UTxO entries that are unsuitable for use as collateral.
- The number of entries within `largestCombinationAvailable` is not greater than `maximumCollateralInputCount`.
- The total value of entries within `largestCombinationAvailable` is not greater than the computed `minimumSelectionAmount` (as this would mean that the error was returned erroneously: we **_could_** have generated a selection).

### Collateral suitability

If we introduce a mutation that deliberately removes the collateral suitability check:
```patch
diff --git a/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs b/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs
index 994103a68..a137cfc05 100644
--- a/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs
+++ b/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs
@@ -316,9 +316,8 @@ toCollateralConstraintsParams balanceResult (constraints, params) =
         }
     collateralParams = Collateral.SelectionParams
         { coinsAvailable =
-            Map.mapMaybeWithKey
-                (curry (view #utxoSuitableForCollateral constraints))
-                (unUTxO (view #utxoAvailableForCollateral params))
+            view (#tokens . #coin) <$> unUTxO
+                (view #utxoAvailableForCollateral params)
         , minimumSelectionAmount =
             computeMinimumCollateral ComputeMinimumCollateralParams
                 { minimumCollateralPercentage =
```
Then we see a verification failure similar to:

collateralErrorVerificationFailure

### Maximum number of collateral inputs

If we introduce a mutation that deliberately passes an incorrect maximum number of collateral inputs to the collateral selection algorithm:
```patch
diff --git a/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs b/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs
index 6fc4d4f5d..1b18e0360 100644
--- a/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs
+++ b/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs
@@ -305,7 +305,7 @@ toCollateralConstraintsParams balanceResult (constraints, params) =
   where
     collateralConstraints = Collateral.SelectionConstraints
         { maximumSelectionSize =
-            view #maximumCollateralInputCount constraints
+            view #maximumCollateralInputCount constraints + 1
         , searchSpaceLimit =
             -- We use the default search space limit here, as this value is
             -- used in the test suite for 'Collateral.performSelection'. We
```

Then we see a verification failure similar to:

collateralErrorVerificationFailure2

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs [hunk 1]
```diff
@@ -745,19 +748,86 @@ verifySelectionLimitReachedError cs ps e
...
+        -- resolved according to the set of UTxOs originally made available
+        -- to the selection algorithm.
+        largestCombinationResolved :: Map TxIn TxOut
+        largestCombinationResolved = Map.restrictKeys
+            (unUTxO (ps ^. #utxoAvailableForCollateral))
+            (Map.keysSet largestCombination)
+
+        -- The subset of the largest reported combination that is suitable for
+        -- use as collateral. This set should be exactly the same size as the
+        -- reporte largest combination.
```
Comment (sevanspowell) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs]:
```suggestion
        -- reported largest combination.
```

**Review by sevanspowell — APPROVED** [review]

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs]: Haha, thanks!

**Pull request merged**